typing ak.Array for numba.cuda.jit signature

### Version of Awkward Array

2.6.2

### Description and code to reproduce

Hey guys, I followed a hint from the discussion in [#696](https://github.com/scikit-hep/awkward/discussions/696#discussion-2571850) to type `ak.Array` for numba signatures. So I tried something like

```python
import awkward as ak
import numba as nb
from numba import types

cpu_arr_type = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend='cpu').numba_type

@nb.njit(types.void(cpu_arr_type))
def cpu_kernel(arr):
   do_something_with_arr
```
and this works like a charm.

However, I'm interested in the same case but with a cuda kernel. So I tried what appeared more natural to do:
```python
gpu_arr_type = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend='cuda').numba_type

@nb.cuda.jit(types.void(gpu_arr_type), extensions=[ak.numba.cuda])
def cuda_kernel(arr):
   do_something_with_arr
```
This time, I get the error:
```python
self = <awkward._connect.numba.arrayview_cuda.ArrayViewArgHandler object at 0x784afbc13fa0>
ty = ak.ArrayView(ak.ListArrayType(array(int64, 1d, C), ak.ListArrayType(array(int64, 1d, C), ak.NumpyArrayType(array(int64, 1d, C), {}), {}), {}), None, ())
val = <Array [[[4, 1], [2, -1]], [...], [[4, 0]]] type='3 * var * var * int64'>
stream = 0, retr = []

    def prepare_args(self, ty, val, stream, retr):
        if isinstance(val, ak.Array):
            if isinstance(val.layout.backend, CupyBackend):
                # Use uint64 for pos, start, stop, the array pointers values, and the pylookup value
                tys = numba.types.UniTuple(numba.types.uint64, 5)
    
>               start = val._numbaview.start
E               AttributeError: 'NoneType' object has no attribute 'start'

.../site-packages/awkward/_connect/numba/arrayview_cuda.py:21: AttributeError
```
How should this latter case be correctly treated? Note that, without typing, the thing works as expected:
```python
@nb.cuda.jit(extensions=[ak.numba.cuda])
def cuda_kernel_no_typing(arr):
   do_something_with_arr
```
However, I'm interested in `ak.Array`s with the 3D layout of integers (as above) and would like to take advantage of numba's eager compilation. I'm passing the `arr` for testing as
```python
backend = 'cpu'  # or 'cuda'
arr = ak.to_backend(
    ak.Array([
        [[4, 1], [2, -1]],
        [[0, -1], [1, 1], [3, -1]],
        [[4, 0]]
    ]),
    backend
)
```
Any help is appreciated!

Thanks for trying @ianna. You're almost there. Up to that point, it also works for me. Now test it with a different array with the same nested structure:
```python
array2 = ak.Array([[[4, 1], [2, -1]], [[0, -1], [1, 1], [3, -1]], [[4, 0]]], backend='cuda')

>>> cuda_kernel[1024, 1024](array2)

../../miniforge3/envs/qb/lib/python3.10/site-packages/numba/cuda/dispatcher.py:539: in __call__
    return self.dispatcher.call(args, self.griddim, self.blockdim,
../../miniforge3/envs/qb/lib/python3.10/site-packages/numba/cuda/dispatcher.py:683: in call
    kernel.launch(args, griddim, blockdim, stream, sharedmem)
../../miniforge3/envs/qb/lib/python3.10/site-packages/numba/cuda/dispatcher.py:317: in launch
    self._prepare_args(t, v, stream, retr, kernelargs)
../../miniforge3/envs/qb/lib/python3.10/site-packages/numba/cuda/dispatcher.py:380: in _prepare_args
    ty, val = extension.prepare_args(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <awkward._connect.numba.arrayview_cuda.ArrayViewArgHandler object at 0x7406bcc4c160>
ty = ak.ArrayView(ak.ListArrayType(array(int64, 1d, C), ak.ListArrayType(array(int64, 1d, C), ak.NumpyArrayType(array(int64, 1d, C), {}), {}), {}), None, ())
val = <Array [[[4, 1], [2, -1]], [...], [[4, 0]]] type='3 * var * var * int64'>
stream = 0, retr = []

    def prepare_args(self, ty, val, stream, retr):
        if isinstance(val, ak.Array):
            if isinstance(val.layout.backend, CupyBackend):
                # Use uint64 for pos, start, stop, the array pointers values, and the pylookup value
                tys = numba.types.UniTuple(numba.types.uint64, 5)
    
>               start = val._numbaview.start
E               AttributeError: 'NoneType' object has no attribute 'start'

../../miniforge3/envs/qb/lib/python3.10/site-packages/awkward/_connect/numba/arrayview_cuda.py:21: AttributeError
```

--- a/src/awkward/_connect/numba/arrayview_cuda.py
+++ b/src/awkward/_connect/numba/arrayview_cuda.py
@@ -17,11 +17,13 @@ class ArrayViewArgHandler:
             if isinstance(val.layout.backend, CupyBackend):
                 # Use uint64 for pos, start, stop, the array pointers values, and the pylookup value
                 tys = numba.types.UniTuple(numba.types.uint64, 5)
-
-                start = val._numbaview.start
-                stop = val._numbaview.stop
-                pos = val._numbaview.pos
-                arrayptrs = val._numbaview.lookup.arrayptrs.data.ptr
+                
+                view = val._numbaview
+                
+                start = view.start
+                stop = view.stop
+                pos = view.pos
+                arrayptrs = view.lookup.arrayptrs.data.ptr
                 pylookup = 0
 
                 return tys, (pos, start, stop, arrayptrs, pylookup)

--- a/src/awkward/_connect/numba/arrayview_cuda.py
+++ b/src/awkward/_connect/numba/arrayview_cuda.py
@@ -20,7 +20,7 @@ class ArrayViewArgHandler:
                         f"the array type: {val.numba_type} does not match "
                         f"the kernel signature type: {ty}"
                     )
-                
+
                 # Use uint64 for pos, start, stop, the array pointers values, and the pylookup value
                 tys = numba.types.UniTuple(numba.types.uint64, 5)
 

--- a/src/awkward/_connect/numba/arrayview_cuda.py
+++ b/src/awkward/_connect/numba/arrayview_cuda.py
@@ -17,9 +17,9 @@ class ArrayViewArgHandler:
             if isinstance(val.layout.backend, CupyBackend):
                 # Use uint64 for pos, start, stop, the array pointers values, and the pylookup value
                 tys = numba.types.UniTuple(numba.types.uint64, 5)
-                
+
                 view = val._numbaview
-                
+
                 start = view.start
                 stop = view.stop
                 pos = view.pos

--- a/src/awkward/_connect/numba/arrayview_cuda.py
+++ b/src/awkward/_connect/numba/arrayview_cuda.py
@@ -15,11 +15,14 @@ class ArrayViewArgHandler:
     def prepare_args(self, ty, val, stream, retr):
         if isinstance(val, ak.Array):
             if isinstance(val.layout.backend, CupyBackend):
+                if ty is not val.numba_type:
+                    raise NumbaTypeError(
+                        f"the array type: {val.numba_type} does not match "
+                        f"the kernel signature type: {ty}"
+                    )
+                
                 # Use uint64 for pos, start, stop, the array pointers values, and the pylookup value
                 tys = numba.types.UniTuple(numba.types.uint64, 5)
-
-                if val._numbaview is None:
-                    _ = val.numba_type
 
                 view = val._numbaview
                 assert view is not None

--- a/src/awkward/_connect/numba/arrayview_cuda.py
+++ b/src/awkward/_connect/numba/arrayview_cuda.py
@@ -17,9 +17,13 @@ class ArrayViewArgHandler:
             if isinstance(val.layout.backend, CupyBackend):
                 # Use uint64 for pos, start, stop, the array pointers values, and the pylookup value
                 tys = numba.types.UniTuple(numba.types.uint64, 5)
+                
+                if val._numbaview is None:
+                    _ = val.numba_type
 
                 view = val._numbaview
-
+                assert(view is not None)
+                
                 start = view.start
                 stop = view.stop
                 pos = view.pos

--- a/src/awkward/_connect/numba/arrayview_cuda.py
+++ b/src/awkward/_connect/numba/arrayview_cuda.py
@@ -17,13 +17,13 @@ class ArrayViewArgHandler:
             if isinstance(val.layout.backend, CupyBackend):
                 # Use uint64 for pos, start, stop, the array pointers values, and the pylookup value
                 tys = numba.types.UniTuple(numba.types.uint64, 5)
-                
+
                 if val._numbaview is None:
                     _ = val.numba_type
 
                 view = val._numbaview
-                assert(view is not None)
-                
+                assert view is not None
+
                 start = view.start
                 stop = view.stop
                 pos = view.pos

--- a/tests-cuda/test_3115_array_typed_cuda_jit.py
+++ b/tests-cuda/test_3115_array_typed_cuda_jit.py
@@ -9,7 +9,7 @@ import awkward as ak
 nb = pytest.importorskip("numba")
 nb_cuda = pytest.importorskip("numba.cuda")
 
-from numba import types
+from numba import types # noqa: E402
 
 nb.config.CUDA_LOW_OCCUPANCY_WARNINGS = False
 nb.config.CUDA_WARN_ON_IMPLICIT_COPY = False

--- a/tests-cuda/test_3115_array_typed_cuda_jit.py
+++ b/tests-cuda/test_3115_array_typed_cuda_jit.py
@@ -2,7 +2,6 @@
 
 from __future__ import annotations
 
-import numpy as np
 import pytest
 
 import awkward as ak
@@ -24,11 +23,11 @@ except ImportError:
 
 def test_array_typed():
     # create an ak.Array with a cuda backend:
-    gpu_arr_type = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend='cuda').numba_type
-    
+    gpu_arr_type = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend="cuda").numba_type
+
     @nb.cuda.jit(types.void(gpu_arr_type), extensions=[ak.numba.cuda])
     def cuda_kernel(arr):
         return None
-    
-    array = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend='cuda')
+
+    array = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend="cuda")
     cuda_kernel[1024, 1024](array)

--- a/tests-cuda/test_3115_array_typed_cuda_jit.py
+++ b/tests-cuda/test_3115_array_typed_cuda_jit.py
@@ -9,7 +9,7 @@ import awkward as ak
 nb = pytest.importorskip("numba")
 nb_cuda = pytest.importorskip("numba.cuda")
 
-from numba import types # noqa: E402
+from numba import types  # noqa: E402
 
 nb.config.CUDA_LOW_OCCUPANCY_WARNINGS = False
 nb.config.CUDA_WARN_ON_IMPLICIT_COPY = False

--- a/tests-cuda/test_3115_array_typed_cuda_jit.py
+++ b/tests-cuda/test_3115_array_typed_cuda_jit.py
@@ -31,3 +31,10 @@ def test_array_typed():
 
     array = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend="cuda")
     cuda_kernel[1024, 1024](array)
+
+    other_array = ak.Array([[1.1, 2.2, 3.3], [None, 4.4], None, [None, 5.5]], backend='cuda')
+    gpu_other_arr_type = other_array.numba_type
+
+    with pytest.raises(nb.core.errors.NumbaTypeError):
+        cuda_kernel[1024, 1024](other_array)
+

--- a/tests-cuda/test_3115_array_typed_cuda_jit.py
+++ b/tests-cuda/test_3115_array_typed_cuda_jit.py
@@ -32,9 +32,10 @@ def test_array_typed():
     array = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend="cuda")
     cuda_kernel[1024, 1024](array)
 
-    other_array = ak.Array([[1.1, 2.2, 3.3], [None, 4.4], None, [None, 5.5]], backend='cuda')
+    other_array = ak.Array(
+        [[1.1, 2.2, 3.3], [None, 4.4], None, [None, 5.5]], backend="cuda"
+    )
     gpu_other_arr_type = other_array.numba_type
 
     with pytest.raises(nb.core.errors.NumbaTypeError):
         cuda_kernel[1024, 1024](other_array)
-

--- a/tests-cuda/test_3115_array_typed_cuda_jit.py
+++ b/tests-cuda/test_3115_array_typed_cuda_jit.py
@@ -35,7 +35,6 @@ def test_array_typed():
     other_array = ak.Array(
         [[1.1, 2.2, 3.3], [None, 4.4], None, [None, 5.5]], backend="cuda"
     )
-    gpu_other_arr_type = other_array.numba_type
 
     with pytest.raises(nb.core.errors.NumbaTypeError):
         cuda_kernel[1024, 1024](other_array)

--- a/tests-cuda/test_3115_array_typed_cuda_jit.py
+++ b/tests-cuda/test_3115_array_typed_cuda_jit.py
@@ -1,0 +1,34 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward/blob/main/LICENSE
+
+from __future__ import annotations
+
+import numpy as np
+import pytest
+
+import awkward as ak
+
+nb = pytest.importorskip("numba")
+nb_cuda = pytest.importorskip("numba.cuda")
+
+from numba import types
+
+nb.config.CUDA_LOW_OCCUPANCY_WARNINGS = False
+nb.config.CUDA_WARN_ON_IMPLICIT_COPY = False
+
+
+try:
+    ak.numba.register_and_check()
+except ImportError:
+    pytest.skip(reason="too old Numba version", allow_module_level=True)
+
+
+def test_array_typed():
+    # create an ak.Array with a cuda backend:
+    gpu_arr_type = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend='cuda').numba_type
+    
+    @nb.cuda.jit(types.void(gpu_arr_type), extensions=[ak.numba.cuda])
+    def cuda_kernel(arr):
+        return None
+    
+    array = ak.Array([[[0, 1], [2, 3]], [[4, 5]]], backend='cuda')
+    cuda_kernel[1024, 1024](array)

